File: backend/src/main/java/com/mailapp/service/EmailService.java
package com.mailapp.service;

import com.mailapp.dto.EmailDto;
import com.mailapp.dto.ReplyEmailRequest;
import com.mailapp.dto.SendEmailRequest;

import java.util.List;

/**
 * Email service contract.
 *
 * Primary implementation: GmailApiServiceImpl
 *
 * EmailController depends ONLY on this interface — not on any concrete provider.
 * The mock implementation remains available for isolated development use,
 * but real API requests are handled by the Gmail implementation.
 */
public interface EmailService {

    /**
     * Returns all emails across all folders.
     * Used to populate the full EmailContext on the frontend.
     */
    List<EmailDto> getAllEmails();

    /**
     * Returns emails in the inbox folder.
     */
    List<EmailDto> getInboxEmails();

    /**
     * Returns sent emails.
     */
    List<EmailDto> getSentEmails();

    /**
     * Returns a single email by its ID.
     * @throws com.mailapp.exception.ResourceNotFoundException if not found
     */
    EmailDto getEmailById(String id);

    String sendEmail(SendEmailRequest request);

    String sendReply(String id, ReplyEmailRequest request);

    void markAsRead(String id);

    void markAsUnread(String id);

    void toggleStar(String id);

    void toggleImportant(String id);

    void moveToTrash(String id);
}

File: frontend/src/services/emailService.ts
/**
 * Email API service.
 *
 * The ONLY module that knows about email-related API endpoints.
 * React components and contexts call these functions — never fetch() directly.
 *
 * Data flow:
 *   Component / Context → emailService → api (base client) → Spring Boot Backend
 *
 * Future: when Gmail is integrated, only the backend changes.
 * This service and the frontend types remain stable.
 */

import { api } from './api';
import { Email } from '../types/email';

/**
 * Fetch all emails (all folders combined).
 * Used by EmailContext to load the initial email list.
 */
export async function getAllEmails(): Promise<Email[]> {
  const response = await api.get<Email[]>('/api/emails');
  return response.data;
}

/**
 * Fetch inbox emails only.
 */
export async function getInboxEmails(): Promise<Email[]> {
  const response = await api.get<Email[]>('/api/emails/inbox');
  return response.data;
}

/**
 * Fetch sent emails only.
 */
export async function getSentEmails(): Promise<Email[]> {
  const response = await api.get<Email[]>('/api/emails/sent');
  return response.data;
}

/**
 * Fetch a single email by ID with full thread.
 */
export async function getEmailById(id: string): Promise<Email> {
  const response = await api.get<Email>(`/api/emails/${id}`);
  return response.data;
}

export interface SendEmailRequest {
  to: string;
  subject: string;
  body: string;
  cc?: string;
  bcc?: string;
}

export async function sendEmail(request: SendEmailRequest): Promise<string> {
  const response = await api.post<string>('/api/emails/send', request);
  return response.data;
}

export async function sendReply(id: string, body: string): Promise<string> {
  const response = await api.post<string>(`/api/emails/${id}/reply`, { body });
  return response.data;
}

export async function markEmailAsRead(id: string): Promise<void> {
  await api.patch<void>(`/api/emails/${id}/read`, {});
}

export async function markEmailAsUnread(id: string): Promise<void> {
  await api.patch<void>(`/api/emails/${id}/unread`, {});
}

export async function toggleEmailStar(id: string): Promise<void> {
  await api.patch<void>(`/api/emails/${id}/star`, {});
}

export async function toggleEmailImportant(id: string): Promise<void> {
  await api.patch<void>(`/api/emails/${id}/important`, {});
}

export async function deleteEmail(id: string): Promise<void> {
  await api.delete<void>(`/api/emails/${id}`);
}

/**
 * Check if the backend is reachable.
 * Returns true if backend is up, false otherwise.
 */
export async function checkBackendHealth(): Promise<boolean> {
  try {
    await api.get('/api/health');
    return true;
  } catch {
    return false;
  }
}
